exec-unittests

```diff
@@ -0,0 +1,90 @@
+# tox environment test execution.
+
+# For now we ignore the tox environment directory argument.
+_toxenvdir="${1}"
+shift
+
+# The first argument will be which major sub-system to run tests for: the
+# agent side of the server side of the source tree.
+
+major="${1}"
+shift
+if [[ -n "${major}" ]]; then
+    if [[ "${major}" != "agent" && "${major}" != "server" ]]; then
+        printf -- "Expected major sub-system to be either 'agent' or 'server', got '%s'\n" "${major}" >&2
+        exit 2
+    fi
+    major_list="${major}"
+else
+    major_list="agent server"
+fi
+
+# Sub-test of the major test.
+subtst="${1}"
+shift
+
+# Remaining positional arguments past along to whatever test is run.
+posargs="${@}"
+
+function run_legacy {
+    printf -- "\n\n\nRunning %s/%s legacy unit tests\n\n" "${1}" "${2}"
+    ${1}/${2}/unittests ${posargs}
+    local _rc=${?}
+    if [[ ${_rc} -ne 0 ]]; then
+        printf -- "\n%s %s legacy tests failed with '%s'\n\n" "${1^}" "${2}" "${_rc}"
+    else
+        printf -- "\n%s %s legacy tests succeeded\n\n" "${1^}" "${2}"
+    fi
+    return ${_rc}
+}
+
+rc=0
+
+if [[ -z "${subtst}" || "${subtst}" == "python" ]]; then
+    _pytest_majors="./lib/pbench/test/unit/common"
+    for _major in ${major_list}; do
+        _pytest_majors="${_pytest_majors} ./lib/pbench/test/unit/${_major}"
+        if [[ "${_major}" == "agent" ]]; then
+            _pytest_majors="${_pytest_majors} ./lib/pbench/test/functional"
+        fi
+    done
+
+    printf -- "\n\n\nRunning %s python3-based unit tests via pytest\n\n" "${major_list}"
+    pytest \
+        --cov --cov-append --cov-report html:cov/html \
+        ${posargs} ${_pytest_majors} || rc=1
+    if [[ ${rc} -ne 0 ]]; then
+        printf -- "\n%s pytest command failed with '%s'\n\n" "${_major^}" "${rc}"
+    else
+        printf -- "\n%s pytest command succeeded\n\n" "${_major^}"
+    fi
+fi
+
+for _major in ${major_list}; do
+    if [[ "${_major}" == "agent" ]]; then
+        if [[ -z "${subtst}" || "${subtst}" == "datalog" ]]; then
+            run_legacy ${_major} tool-scripts/datalog ${posargs} || rc=1
+        fi
+        if [[ -z "${subtst}" || "${subtst}" == "postprocess" ]]; then
+            run_legacy ${_major} tool-scripts/postprocess ${posargs} || rc=1
+        fi
+        if [[ -z "${subtst}" || "${subtst}" == "tool-scripts" ]]; then
+            run_legacy ${_major} tool-scripts ${posargs} || rc=1
+        fi
+        if [[ -z "${subtst}" || "${subtst}" == "util-scripts" ]]; then
+            run_legacy ${_major} util-scripts ${posargs} || rc=1
+        fi
+        if [[ -z "${subtst}" || "${subtst}" == "bench-scripts" ]]; then
+            run_legacy ${_major} bench-scripts ${posargs} || rc=1
+        fi
+    elif [[ "${_major}" == "server" ]]; then
+        if [[ -z "${subtst}" || "${subtst}" == "legacy" ]]; then
+            run_legacy ${_major} bin ${posargs} || rc=1
+        fi
```

Comment:
Since every one of the `run_legacy` command lines is unique anyway, we might as well replace the `${major}` with the value that it's going to have -- that will be clearer for the reader and it will prevent oopsies if the code gets disturbed.

(I wish there were a nice way to catch typos in the `${subtst}` value, but you can't do `elif`'s here because you want them _all_ to run by default.  🙁)

Reply:
Fixing.